backend/mpmediabridge.m

```diff
@@ -0,0 +1,85 @@
+//go:build darwin
+
+#import "mpmediabridge.h"
+
+/**
+ * C bridge registering callbacks for media playback events using the native CommandCenter API.
+ */
+void register_os_remote_commands() {
+    MPRemoteCommandCenter *commandCenter = [MPRemoteCommandCenter sharedCommandCenter];
+    [commandCenter.playCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent * _Nonnull event) {
+        os_remote_command_callback(PLAY);
+        return MPRemoteCommandHandlerStatusSuccess;
+    }];
+
+    [commandCenter.pauseCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent * _Nonnull event) {
+        os_remote_command_callback(PAUSE);
+        return MPRemoteCommandHandlerStatusSuccess;
+    }];
+
+    [commandCenter.togglePlayPauseCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent * _Nonnull event) {
+        os_remote_command_callback(TOGGLE);
+        return MPRemoteCommandHandlerStatusSuccess;
+    }];
+
+    [commandCenter.stopCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent * _Nonnull event) {
+        os_remote_command_callback(STOP);
+        return MPRemoteCommandHandlerStatusSuccess;
+    }];
+
+    [commandCenter.nextTrackCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent * _Nonnull event) {
+        os_remote_command_callback(NEXT_TRACK);
+        return MPRemoteCommandHandlerStatusSuccess;
+    }];
+
+    [commandCenter.previousTrackCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent * _Nonnull event) {
+        os_remote_command_callback(PREVIOUS_TRACK);
+        return MPRemoteCommandHandlerStatusSuccess;
+    }];
+}
+
+/**
+ * C bridge setting "Now Playing" information on macOS for media playback using the native APIs.
+ */
+void set_os_now_playing_info(const char *title, const char *artist, const char *coverArtFileURL) {
+    NSString *coverArtLocationString = [NSString stringWithUTF8String:coverArtFileURL];
+    NSURL *coverArtURL = [NSURL URLWithString:coverArtLocationString];
```

Comment:
This might be unnecessary optimization, but I wonder if the OS will keep the previous image if you don't re-set it in the NSDictionary when changing other metadata? If so, you could store the previous image URL as a global var and check if the URL on the next call is the same and not need to re-set the image, for example if you are seeking through tracks from the same album.

Again this may be premature optimization not worth the effort, so up to you.

Reply:
In this case, not setting the key will result in no image being displayed.

Also; remember this data is a read from cache. So if implemented, even if we stored the image itself in-memory we would be optimising for a single disk read (and potentially using more memory)